tempesta_fw/apm.c

```diff
@@ -0,0 +1,1041 @@
+	jtmistart = jtmnow - (jtmnow % tfw_apm_jtmintrvl);
+	/* The start of the current time window. */
+	jtmwstart = jtmistart - tfw_apm_jtmwindow;
+	/* The index of the current entry. */
+	centry = (jtmnow / tfw_apm_jtmintrvl) % rbuf->rbufsz;
+
+	/*
+	 * If the latest percentiles are for a different time window,
+	 * then a recalculation is in order.
+	 */
+	if (unlikely(rbctl->jtmwstamp != jtmwstart)) {
+		tfw_apm_rbent_checkreset(&rbent[centry], jtmistart);
+
+		for (i = 0; i < rbuf->rbufsz; ++i)
+			total_cnt += atomic64_read(&rbent[i].pcntrng.total_cnt);
+
+		rbctl->entry_cnt = 0;
+		rbctl->total_cnt = total_cnt;
+		rbctl->jtmwstamp = jtmwstart;
+
+		TFW_DBG3("%s: New time window: centry [%d] total_cnt [%lu]\n",
+			 __func__, centry, rbctl->total_cnt);
+
+		return true;
+	}
+
+	/* The latest percentiles are for the current time window.
+	 * In some cases a recalculation is not required. In some
+	 * other cases the recalculation set up can be simpler.
+	 */
+
+	/* Nothing to do if there were no stats updates. */
+	entry_cnt = atomic64_read(&rbent[centry].pcntrng.total_cnt);
+	if (unlikely(rbctl->entry_cnt == entry_cnt)) {
+		if (unlikely(recalc)) {
+			TFW_DBG3("%s: Old time window: recalculate: "
+				 "centry [%d] total_cnt [%lu]\n",
+				 __func__, centry, rbctl->total_cnt);
+			TFW_INC_STAT_BH(apm.calc_repeats);
+			return true;
+		}
+		return false;
+	}
+	BUG_ON(rbctl->entry_cnt > entry_cnt);
+
+	/* Update the counts incrementally. */
+	rbctl->total_cnt += entry_cnt - rbctl->entry_cnt;
+	rbctl->entry_cnt = entry_cnt;
+
+	TFW_DBG3("%s: Old time window: centry [%d] total_cnt [%lu]\n",
+		 __func__, centry, rbctl->total_cnt);
+
+	return true;
+}
+
+/*
+ * Calculate the latest percentiles if necessary.
+ *
+ * Return the number of percentile values that have been filled
+ * if potentially new percentile values were calculated.
+ * Return 0 if the percentile values didn't need the recalculation.
+ * REturn -1 of the recalculation could not be performed.
+ */
+static int
+__tfw_apm_calc(TfwApmData *data, TfwApmPrcntl *prcntl, int recalc)
+{
+	int ret;
+
+	if (!spin_trylock(&tfw_sa_guard)) {
+		TFW_INC_STAT_BH(apm.calc_misses);
+		return -1;
+	}
+	if ((ret = tfw_apm_rbctl_update(data, recalc)))
+		ret = tfw_apm_prnctl_calc(&data->rbuf, &data->rbctl, prcntl);
+	spin_unlock(&tfw_sa_guard);
+
+	return ret;
+}
+
+/*
+ * Calculate the latest percentiles if necessary.
+ *
+ * Note that this function may also be used concurrently by other users
+ * than the kernel timer function in this module, should the need arise.
+ * That should only be done in exceptional cases (like testing), because
+ * it would increase @tfw_sa_guard lock contention.
+ */
+static void
+tfw_apm_calc(TfwApmData *data)
+{
+	int nfilled, wridx, recalc;
+	PrcntlStats pstats[ARRAY_SIZE(tfw_apm_pstats)];
+	TfwApmPrcntl wrkprcntl = { pstats, ARRAY_SIZE(pstats) };
+	TfwApmPrcntl *prcntl;
+
+	memcpy(pstats, tfw_apm_pstats, sizeof(tfw_apm_pstats));
+
+	smp_mb__before_atomic();
```

Comment:
In my understanding, this is not a compiler barrier on an SMP system, but a memory barrier for synchronization between CPUs. `atomic_read()` doesn't imply memory barriers. I would want to be absolutely sure that this is the entry for reading the latest calculated percentiles, as the index may be incremented on another CPU.

On the other hand, there's also `test_and_clear_bit()` here that does imply a memory barrier. Perhaps, we can put that statement first and get the memory barrier for `atomic_read()` for free? That's a bit non-obvious though.

**ADDED:** Well, as it turns out `rdidx` is not needed that much, then this barrier is automatically not needed as well.